keps/sig-node/20190604-rootless.md

```diff
@@ -0,0 +1,213 @@
+
+However, these are not efficient to mitigate vulnerabilities of the node components themselves (kubelet, CRI, OCI...).
+
+e.g.
+
+- [CVE-2017-1002102](https://nvd.nist.gov/vuln/detail/CVE-2017-1002102): kubelet could delete files on the host during syncing secret/configMap/downwardAPI volumes
+- [CVE-2019-11245](https://nvd.nist.gov/vuln/detail/CVE-2019-11245): Dockerfile USER instruction was ignored
+- [CVE-2018-11235](https://nvd.nist.gov/vuln/detail/CVE-2018-11235): kubelet could execute an arbitrary command as the root via gitRepo volumes
+- Potential image extraction [zip-slip](https://snyk.io/research/zip-slip-vulnerability) vulnerabilities in CRI runtimes. Both containerd and CRI-O are working on implementing supports for new archive formats like zstd, imgcrypt, and stargz. Potentially these implementations have such vulnerabilities.
+- And lots of CRI/OCI vulnerabilities in the past.
+
+## Goals
+* Step 1: Allow `kubelet` and `kube-proxy` being executed inside user namespaces create by a non-root user [**mergable**]
+  * See [User namespaces](#user-namespaces) and https://github.com/kubernetes/kubernetes/pull/78635
+* Step 2: Allow `kubelet` to configure cgroup v2 via a user-instance of systemd [**Waiting for cgroup2 KEP to settle down**]
+  * See [cgroup](#cgroup), https://github.com/kubernetes/enhancements/pull/1370
+* Step 3: Allow `kube-proxy` to propagate NodePort ports from its network namespace into the initial network namespace (optional)
+  * See [Network namespaces](#network-namespaces)
+
+## Implementation Details
+
+### User namespaces
+All components need to be executed inside a user namespace (along with a network namespace and a mount namespace)
+to gain fake-root privileges, mostly for network and mount operations.
+
+These namespaces are expected to be created using [RootlessKit](https://github.com/rootless-containers/rootlesskit).
+In short, RootlessKit is an extended version of [`unshare -rmn`](http://man7.org/linux/man-pages/man1/unshare.1.html).
+
+RootlessKit has been already adopted by Docker, BuildKit, Usernetes, and k3s.
+
+> **Note**
+> Rootless mode is unrelated to [the Node-level UserNS KEP](https://github.com/kubernetes/enhancements/issues/127).
+>
+> Rootless mode executes all components inside UserNS to mitigate vulnerabilities of all components,
+> while Node-level UserNS executes only containers inside UserNS.
+>
+> Rootless mode and Node-level UserNS do not conflict and can be stacked together. (Node-level UserNS inside Rootless UserNS.)
+
+#### Paths
+
+Some paths like `/var/log/pods` are hardcoded in Kubernetes and hard to change.
+As RootlessKit can bind-mount writable directories on these paths, Kubernetes does not need to be updated.
+
+#### Pod SecurityContext
+
+- `runAsUser`: supported, but the number of the UID is limited by `/etc/subuid`.
+- `sysctls`: some sysctl parameters are supported, but some would fail in `EPERM`.
```

Comment:
AFAIK many K8s users are using this feature heavily.

Reply:
What kind of sysctl values especially?